ssl/quic/quic_port.c

```diff
@@ -853,39 +1021,88 @@ static void port_send_version_negotiation(QUIC_PORT *port, BIO_ADDR *peer,
+ * parameters. If the token is from a NEW_TOKEN packet, the values will be
+ * derived instead.
  *
  * @param hdr   Pointer to the QUIC packet header containing the token.
- * @param odcid Pointer to the connection ID structure to store the ODCID if
- *              the token is valid.
- * @return      1 if the token is valid and ODCID is extracted successfully,
+ * @param port  Pointer to the QUIC port from which to send the packet.
+ * @param peer  Address of the client peer receiving the packet.
+ * @param odcid Pointer to the connection ID structure to store the ODCID if the
+ *              token is valid.
+ * @param scid  Pointer to the connection ID structure to store the SCID if the
+ *              token is valid.
+ *
+ * @return      1 if the token is valid and ODCID/SCID are successfully set.
  *              0 otherwise.
  *
  * The function performs the following checks:
- * - Verifies that the token length meets the required minimum.
- * - Confirms the token buffer matches the expected "openssltoken" string.
- * -
+ * - Token length meets the required minimum.
+ * - Buffer matches expected format.
+ * - Peer address matches previous connection address.
+ * - Token has not expired. Currently set to 10 seconds for tokens from RETRY
+ *   packets and 60 minutes for tokens from NEW_TOKEN packets. This may be
+ *   configurable in the future.
  */
-static int port_validate_token(QUIC_PKT_HDR *hdr, QUIC_CONN_ID *odcid)
+static int port_validate_token(QUIC_PKT_HDR *hdr, QUIC_PORT *port,
+                               BIO_ADDR *peer, QUIC_CONN_ID *odcid,
+                               QUIC_CONN_ID *scid)
 {
-    int valid;
-    struct validation_token *token;
+    int ret = 0;
+    QUIC_VALIDATION_TOKEN token = { 0 };
+    unsigned long long time_diff;
+    size_t remote_addr_len;
+    unsigned char *remote_addr = NULL;
 
-    memset(odcid, 0, sizeof(QUIC_CONN_ID));
+    if (!parse_validation_token(&token, hdr->token, hdr->token_len))
+        goto err;
 
-    token = (struct validation_token *)hdr->token;
-    if (token == NULL || hdr->token_len <= (TOKEN_LEN - QUIC_RETRY_INTEGRITY_TAG_LEN))
-        return 0;
+    /* Validate token timestamp */
+    time_diff = ossl_time2seconds(ossl_time_abs_difference(token.timestamp,
+                                                           ossl_time_now()));
+    if ((token.is_retry && time_diff > 10)
+        || (!token.is_retry && time_diff > 3600))
+        goto err;
 
-    valid = memcmp(token->token_buf, "openssltoken", sizeof("openssltoken") - 1);
-    if (valid != 0)
-        return 0;
+    /* Validate remote address */
+    if (!BIO_ADDR_rawaddress(peer, NULL, &remote_addr_len)
+        || remote_addr_len != token.remote_addr_len
+        || (remote_addr = OPENSSL_malloc(remote_addr_len)) == NULL
+        || !BIO_ADDR_rawaddress(peer, remote_addr, &remote_addr_len)
+        || CRYPTO_memcmp(remote_addr, token.remote_addr, remote_addr_len) != 0)
```

Comment:
Just memcmp seems enough. `CRYPTO_memcmp` is for constant time behaviour, which seems overkill here.